### ISSUE TITLE
【第五章】DCGANの学習について

### ISSUE THREAD

**Issue by Yoshi-0921** [YutaroOgawa/pytorch_advanced#86]
P.259から説明されているDCGANの学習コードについて質問がございます。

train_model関数内の最後でEpochごとのLossを計算する（P.262）と記載されていますが、
何故この場合はepoch_d_loss/num_train_imgsではなく、epoch_d_loss/batch_sizeとなるのでしょうか。（epoch_g_lossの計算でも同様です）私としては、損失の合計値をデータの個数分で割ることによって１Epoch内で得た損失の平均値を算出していると認識していたのですが、DCGANの場合では損失値の計算方法が異なってくるのでしょうか。

拙い質問文で申し訳ございません。
どうぞよろしくお願い致します。

**Comment by YutaroOgawa** [comment]
@Yoshi-0921 さま

ご質問ありがとうございます。
該当箇所は、

```python

print('epoch {} || Epoch_D_Loss:{:.4f} ||Epoch_G_Loss:{:.4f}'.format(
epoch, epoch_d_loss/batch_size, epoch_g_loss/batch_size))
```

と認識しました。
おっしゃる通り、非常にちぐはぐな状況になっています。
混乱と困惑を招いてしまい、大変申し訳ございません。
コードが良くないです。


p. 260あたりより、実装を一部分ずつ、抜き出すと

```python
# 誤差関数を定義
criterion = nn.BCEWithLogitsLoss(reduction='mean')

・・・

# 画像の枚数
num_train_imgs = len(dataloader.dataset)
batch_size = dataloader.batch_size

・・・

epoch_d_loss += d_loss.item()

・・・

print('epoch {} || Epoch_D_Loss:{:.4f} ||Epoch_G_Loss:{:.4f}'.format(
epoch, epoch_d_loss/batch_size, epoch_g_loss/batch_size))
```

となっています。

誤差関数はredcution='mean'により、ミニバッチサイズによらず、ミニバッチでの平均の値を求めています。

それを全ミニバッチ分計算したのが、epoch_d_lossやepoch_g_loss

です。

そのため、画像1枚に対する損失を表示するには、
ミニバッチを取り出す回数がおおよそ
num_train_imgs / batch_size [回]
なので、これで、トータルのlosssを割ってあげるべきです。

そのため、以下となります。

```python
print('epoch {} || Epoch_D_Loss:{:.4f} ||Epoch_G_Loss:{:.4f}'.format(
epoch, epoch_d_loss/num_train_imgs*batch_size, epoch_g_loss/num_train_imgs*batch_size))
```


該当部分はprintしているだけなので、処理に影響はないのですが、
困惑を生んでしまう実装になっており、大変申し訳ございません。

どうぞよろしくお願い致します。

**Comment by Yoshi-0921** [comment]
Ogawa様

非常に丁寧なご説明、誠にありがとうございました。より本書での学習が捗りました。

失礼ながらもう一点、お伺いしたい内容がございます。第6-4章のEfficient GANの実装コードについてです。本書の説明では、def train_model下で以下のように損失関数を求めた後、各optimizerで重みを更新しています。

`# 誤差を計算
            d_loss_real = criterion(d_out_real.view(-1), label_real)
            d_loss_fake = criterion(d_out_fake.view(-1), label_fake)
            d_loss = d_loss_real + d_loss_fake

            # バックプロパゲーション
            d_optimizer.zero_grad()
            d_loss.backward()
            d_optimizer.step()`
`# 誤差を計算
            g_loss = criterion(d_out_fake.view(-1), label_real)

            # バックプロパゲーション
            g_optimizer.zero_grad()
            g_loss.backward()
            g_optimizer.step()`
`# 誤差を計算
            e_loss = criterion(d_out_real.view(-1), label_fake)

            # バックプロパゲーション
            e_optimizer.zero_grad()
            e_loss.backward()
            e_optimizer.step()`

上の例では、各optimizerを更新する際、その該当optimizerのみの勾配をリセットしています。
例として、d_optimizer.step()する際、g_optimizer.zero_grad()やe_optimizer.zero_grad()を実行せず、d_optimizer.zero_grad()のみを実行しています。

しかし、本書第五章で説明されていたDCGANの実装では、以下のように両方のoptimizerのバックプロパゲーションを行っています。

`# 誤差を計算
            d_loss_real = criterion(d_out_real.view(-1), label_real)
            d_loss_fake = criterion(d_out_fake.view(-1), label_fake)
            d_loss = d_loss_real + d_loss_fake

            # バックプロパゲーション
            g_optimizer.zero_grad()
            d_optimizer.zero_grad()

            d_loss.backward()
            d_optimizer.step()`
`# 誤差を計算
            g_loss = criterion(d_out_fake.view(-1), label_real)

            # バックプロパゲーション
            g_optimizer.zero_grad()
            d_optimizer.zero_grad()
            g_loss.backward()
            g_optimizer.step()`

このように、Efficient GANでは上の例ように、全てのoptimizerのバックプロパゲーションを毎回求めていない点が疑問に思いました。（あえて勾配をリセットしないことにより、勾配を蓄積させる目的があるのでしょうか。）

何度も質問申し訳ございません。どうぞよろしくお願いいたします。

**Comment by YutaroOgawa** [comment]
@Yoshi-0921 さま

ありがとうございます。

もう1点の質問ですが、

第5章のDCGANにて、loss.backwards()する前に記述している2つのoptimizer.zero_grad()のうち、
loss.backwardsしない方のoptimizer.zero_grad()
が、実は不要な気がしてきました。

実際にコメントアウトして、動作させても、おおむね同じ挙動をしているように見えます。

```python
            d_loss_real = criterion(d_out_real.view(-1), label_real)
            d_loss_fake = criterion(d_out_fake.view(-1), label_fake)
            d_loss = d_loss_real + d_loss_fake
            # バックプロパゲーション
            # 不要か　g_optimizer.zero_grad()
            d_optimizer.zero_grad()
            d_loss.backward()
            d_optimizer.step()

# 誤差を計算
            g_loss = criterion(d_out_fake.view(-1), label_real)
            # バックプロパゲーション
            g_optimizer.zero_grad()
            # 不要か d_optimizer.zero_grad()
            g_loss.backward()
            g_optimizer.step()
```


まず、モデルに画像を入力しアウトプットを計算した時点で、アウトプット値がモデルの変数で示されます。

アウトプットの値を誤差関数でさらに計算することで、lossの値がモデルの変数で示されます。

このlossをloss.backward()したときに、optimizerで対象にしているモデル変数は誤差逆伝搬で勾配値が計算され、optimizerのその変数の勾配値に足し算されます。

そして、optimizer.step()でモデル変数の値が勾配値に基づいて、更新されます。

optimizerには勾配値はどんどん足し算されるので、backward()を計算する前に、溜まっている勾配の値をリセットして0にする必要があり、それが、optimizer.zero_grad()
になります。

GANの場合、2つのネットワーク・モデルを通るので、両方のoptimizerをzero_grad()していましたが、その後、loss.backward()する方だけをzero_grad()で良い気がしてきました。

大変申し訳ございませんが、この回答にはあまり自信を持ちきれないのですが、

第6章のefficient netが正しく、第5章のDCGANは間違ってはいないが不要なoptimizer.zero_grad()がある状態な気がします。

正確なのはPyTorchの公式のDCGANの実装例ですが、これは私とはまた違った実装になっています（Networkをzero_gradしている）。

https://pytorch.org/tutorials/beginner/dcgan_faces_tutorial.html


きちんとした回答になっておらず、大変申し訳ございませんが、上記でお願いいたします。
